### PR TITLE
feat: support hot/cold store in state viewer and add view-trie2 command 

### PR DESCRIPTION
**Pull request by wacban**
- added store_temperature cmdline argument to neard view-state 
- added neard view-state view-trie2 command that is similary to view-trie but
  - can only iterate from the state root node
  - prints leaves only
  - prints data in more human friendly way
- added max_depth to the trie iterator (needed for the view-trie2)
- refactored the retrieve_raw_node_or_value to return NodeOrValue instead of Result

### CODE REVIEW

File: chain/network/src/peer_manager/peer_store/mod.rs [hunk 1]
```diff
@@ -592,12 +594,12 @@ impl PeerStore {
 }
 
 /// Public method used to iterate through all peers stored in the database.
-pub fn iter_peers_from_store<F>(store: near_store::NodeStorage, f: F)
+pub fn iter_peers_from_store<F>(db: Arc<dyn Database>, f: F)
```
Comment (wacban) [inline chain/network/src/peer_manager/peer_store/mod.rs:597]: Note for reviewers:
This is used by the state-view command where I want to select what store (hot/ cold) to use and only pass in the store of chosen temperature.

File: core/store/src/lib.rs [hunk 2]
```diff
@@ -69,6 +70,20 @@ pub enum Temperature {
     Cold,
 }
 
+impl FromStr for Temperature {
```
Comment (wacban) [inline core/store/src/lib.rs:73]: Note for reviewers:
This is used for nice cmdline parsing of temperature.

File: core/store/src/trie/iterator.rs [hunk 3]
```diff
@@ -238,8 +246,8 @@ impl<'a> TrieIterator<'a> {
                 Some(IterStep::Value(hash))
             }
             (CrumbStatus::At, TrieNode::Extension(key, child)) => {
-                let hash = *child.unwrap_hash();
                 let key = NibbleSlice::from_encoded(key).0;
+                let hash = *child.unwrap_hash();
```
Comment (wacban) [inline core/store/src/trie/iterator.rs:250]: Note for self: accidental change, ideally revert that.

File: core/store/src/trie/iterator.rs [hunk 4]
```diff
@@ -367,15 +375,24 @@ impl<'a> Iterator for TrieIterator<'a> {
...
             let iter_step = self.iter_step()?;
+
+            let can_process = match self.max_depth {
+                Some(max_depth) => self.key_nibbles.len() <= max_depth,
+                None => true,
+            };
+            if !can_process {
+                continue;
+            }
+
```
Comment (wacban) [inline core/store/src/trie/iterator.rs:386]: Note for reviewers:
There are many places where this if statement could go but here at least it's only in a single place and handles all the node types. There is minor perf penalty for doing it here - here we already have the nodes exceeding max depth lined up for iteration. I feel this is much better becuase it's cleaner and it's not perf-critical.

File: core/store/src/trie/mod.rs [hunk 5]
```diff
@@ -795,12 +835,12 @@ impl Trie {
     }
 
     // Similar to retrieve_raw_node but handles the case where there is a Value (and not a Node) in the database.
-    fn retrieve_raw_node_or_value(
-        &self,
-        hash: &CryptoHash,
-    ) -> Result<Result<RawTrieNodeWithSize, std::sync::Arc<[u8]>>, StorageError> {
+    fn retrieve_raw_node_or_value(&self, hash: &CryptoHash) -> Result<NodeOrValue, StorageError> {
         let bytes = self.storage.retrieve_raw_bytes(hash)?;
-        Ok(RawTrieNodeWithSize::decode(&bytes).map_err(|_| bytes))
```
Comment (wacban) [inline core/store/src/trie/mod.rs:803]: Note for reviewers:
Sure, map is nice but returning a correct value (NodeValue) in Result::Err is not cool.

File: tools/state-viewer/src/commands.rs [hunk 6]
```diff
@@ -23,12 +23,13 @@ use near_primitives::trie_key::TrieKey;
...
 use near_primitives::types::{BlockHeight, ShardId, StateRoot};
 use near_primitives_core::types::Gas;
+use near_store::db::Database;
 use near_store::test_utils::create_test_store;
+use near_store::Store;
 use near_store::Trie;
 use near_store::TrieCache;
 use near_store::TrieCachingStorage;
 use near_store::TrieConfig;
-use near_store::{NodeStorage, Store};
```
Comment (wacban) [inline tools/state-viewer/src/commands.rs:31]: Note for self: this could be single line.

**Comment by wacban** [comment]
test plan:
- cargo build
- cargo build --features cold_store
- neard view-state view-chain ----> height at ~300 (localnet)
- neard view-state -t hot view-chain ----> height at ~300 (localnet)
- neard view-state -t cold view-chain ----> height at 2 (localnet, prepared manually)
- neard view-state view-trie --hash <hash>  --max-depth 100 --shard-id 0 --shard-version 0 ----> works
- neard view-state view-trie2 --state-root-hash <hash>  --max-depth 100 --shard-id 0 --shard-version 0 ----> works

```
near Some(Account { account_id: AccountId("near"), account: Account { amount: 1000000106545122876440736056756514, locked: 0, code_hash: 11111111111111111111111111111111, storage_usage: 182, version: V1 } })
node0 Some(Account { account_id: AccountId("node0"), account: Account { amount: 950000000000000000000000000000000, locked: 50000958906105887966624510808658, code_hash: 11111111111111111111111111111111, storage_usage: 182, version: V1 } })
                                        Some(AccessKey { account_id: AccountId("near"), public_key: ed25519:546XB2oHhj7PzUKHiH9Xve3Ze5q1JiW2WTh6abXFED3c, access_key: AccessKey { nonce: 0, permission: FullAccess } })
                                         Some(AccessKey { account_id: AccountId("node0"), public_key: ed25519:7PGseFbWxvYVgZ89K1uTJKYoKetWs7BJtbyXDzfbAcqX, access_key: AccessKey { nonce: 0, permission: FullAccess } })
```
- neard view-state view-trie2 --state-root-hash <hash>  --max-depth 10 --shard-id 0 --shard-version 0 ----> works and only prints the accounts from above. Note that depth is measured in nibbles :)

File: core/store/src/trie/iterator.rs [hunk 7]
```diff
@@ -4,13 +4,17 @@ use crate::trie::nibble_slice::NibbleSlice;
 use crate::trie::{TrieNode, TrieNodeWithSize, ValueHandle};
 use crate::{StorageError, Trie};
 
+/// Crumb is a piece of trie iteration state. It describes a node on the trail and processing status of that node.
```
Comment (nikurt) [inline core/store/src/trie/iterator.rs:7]: Thank you :+1:

File: core/store/src/trie/iterator.rs [hunk 8]
```diff
@@ -59,12 +66,13 @@ pub struct TrieTraversalItem {
...
     #![allow(clippy::new_ret_no_self)]
     /// Create a new iterator.
-    pub(super) fn new(trie: &'a Trie) -> Result<Self, StorageError> {
+    pub(super) fn new(trie: &'a Trie, max_depth: Option<usize>) -> Result<Self, StorageError> {
         let mut r = TrieIterator {
             trie,
             trail: Vec::with_capacity(8),
             key_nibbles: Vec::with_capacity(64),
             visited_nodes: None,
+            max_depth: max_depth,
```
Comment (nikurt) [inline core/store/src/trie/iterator.rs:75]:
```suggestion
            max_depth,
```

File: core/store/src/trie/iterator.rs [hunk 9]
```diff
@@ -367,15 +375,24 @@ impl<'a> Iterator for TrieIterator<'a> {
     fn next(&mut self) -> Option<Self::Item> {
         loop {
             let iter_step = self.iter_step()?;
+
+            let can_process = match self.max_depth {
+                Some(max_depth) => self.key_nibbles.len() <= max_depth,
+                None => true,
+            };
+            if !can_process {
+                continue;
```
Comment (nikurt) [inline core/store/src/trie/iterator.rs:384]: I don't understand the code perfectly to check if this works as expected, but

if you do `continue` here, `self.trail.pop();` at line 390 will not happen. Wouldn't it violate some invariants?

File: tools/state-viewer/src/cli.rs [hunk 10]
```diff
@@ -512,3 +532,23 @@ impl ViewTrieCmd {
         view_trie(store, hash, self.shard_id, self.shard_version, self.max_depth).unwrap();
     }
 }
+
+#[derive(Parser)]
+pub struct ViewTrie2Cmd {
```
Comment (nikurt) [inline tools/state-viewer/src/cli.rs:537]: `ViewTrie2` isn't a user-friendly name :)

Taking a step back. Can this be merged with the `ViewTrie` command? For example, as a subcommand to choose between the two behaviors?

File: tools/state-viewer/src/cli.rs [hunk 11]
```diff
@@ -512,3 +532,23 @@ impl ViewTrieCmd {
...
+    #[clap(long)]
+    state_root_hash: String,
+    #[clap(long)]
+    shard_id: u32,
+    #[clap(long)]
+    shard_version: u32,
+    #[clap(long)]
+    max_depth: u32,
```
Comment (nikurt) [inline tools/state-viewer/src/cli.rs:545]: Please comment that setting `max_depth` too high isn't recommended for nodes with large state.

File: tools/state-viewer/src/cli.rs (hunk 10 above)
Comment (wacban) [reply tools/state-viewer/src/cli.rs]: I merged both into a single command and added --format to choose between full or pretty formatting.

File: tools/state-viewer/src/cli.rs [hunk 12]
```diff
@@ -494,21 +509,69 @@ impl ViewChainCmd {
     }
 }
 
+#[derive(Subcommand)]
```
Comment (nikurt) [inline tools/state-viewer/src/cli.rs:512]: Does this need to be a subcommand? a required flag should be enough.

**Review by nikurt — APPROVED** [review]

**Comment by wacban** [comment]
Added --store-temperature to view-state and implemented a new, pretty, output format for view-steate view-trie.

**Pull request merged**